// @angular-generic-table/core/core.module.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var common_1 = require("@angular/common");
var forms_1 = require("@angular/forms");
var generic_table_component_1 = require("./components/generic-table.component");
var gt_render_pipe_1 = require("./pipes/gt-render.pipe");
var gt_visible_pipe_1 = require("./pipes/gt-visible.pipe");
var dash_case_pipe_1 = require("./pipes/dash-case.pipe");
var gt_property_pipe_1 = require("./pipes/gt-property.pipe");
var gt_chunk_pipe_1 = require("./pipes/gt-chunk.pipe");
var gt_filter_pipe_1 = require("./pipes/gt-filter.pipe");
var gt_order_by_pipe_1 = require("./pipes/gt-order-by.pipe");
var gt_expanding_row_component_1 = require("./components/gt-expanding-row.component");
var gt_search_pipe_1 = require("./pipes/gt-search.pipe");
var component_anchor_directive_1 = require("./directives/component-anchor.directive");
var gt_pagination_component_1 = require("./components/gt-pagination.component");
var gt_table_info_component_1 = require("./components/gt-table-info.component");
var gt_custom_component_factory_1 = require("./components/gt-custom-component-factory");
var gt_meta_pipe_1 = require("./pipes/gt-meta.pipe");
var gt_dropdown_component_1 = require("./components/gt-dropdown.component");
var gt_totals_pipe_1 = require("./pipes/gt-totals.pipe");
var gt_totals_position_pipe_1 = require("./pipes/gt-totals-position.pipe");
var gt_row_class_pipe_1 = require("./pipes/gt-row-class.pipe");
var gt_column_class_pipe_1 = require("./pipes/gt-column-class.pipe");
var gt_column_width_directive_1 = require("./directives/gt-column-width.directive");
var gt_drilldown_component_1 = require("./components/gt-drilldown.component");
var GenericTableModule = (function () {
    function GenericTableModule() {
    }
    return GenericTableModule;
}());
GenericTableModule.decorators = [
    { type: core_1.NgModule, args: [{
                declarations: [
                    component_anchor_directive_1.ComponentAnchorDirective,
                    generic_table_component_1.GenericTableComponent,
                    gt_pagination_component_1.GtPaginationComponent,
                    gt_visible_pipe_1.GtVisiblePipe,
                    gt_render_pipe_1.GtRenderPipe,
                    dash_case_pipe_1.DashCasePipe,
                    gt_property_pipe_1.GtPropertyPipe,
                    gt_chunk_pipe_1.GtChunkPipe,
                    gt_filter_pipe_1.GtFilterPipe,
                    gt_order_by_pipe_1.GtOrderByPipe,
                    gt_expanding_row_component_1.GtExpandingRowComponent,
                    gt_custom_component_factory_1.GtCustomComponentFactory,
                    gt_search_pipe_1.GtSearchPipe,
                    gt_pagination_component_1.PaginationPipe,
                    gt_table_info_component_1.GtTableInfoComponent,
                    gt_table_info_component_1.TableInfoPipe,
                    gt_meta_pipe_1.GtMetaPipe,
                    gt_totals_pipe_1.GtTotalsPipe,
                    gt_totals_position_pipe_1.GtTotalsPositionPipe,
                    gt_row_class_pipe_1.GtRowClassPipe,
                    gt_column_class_pipe_1.GtColumnClassPipe,
                    gt_dropdown_component_1.GtDropdownComponent,
                    gt_column_width_directive_1.GtColumnWidthDirective,
                    gt_drilldown_component_1.GtDrilldownComponent
                ],
                imports: [common_1.CommonModule, forms_1.FormsModule],
                exports: [
                    generic_table_component_1.GenericTableComponent,
                    gt_pagination_component_1.GtPaginationComponent,
                    gt_table_info_component_1.GtTableInfoComponent,
                    gt_property_pipe_1.GtPropertyPipe,
                    gt_expanding_row_component_1.GtExpandingRowComponent,
                    gt_dropdown_component_1.GtDropdownComponent,
                    gt_drilldown_component_1.GtDrilldownComponent
                ],
                entryComponents: [
                    gt_drilldown_component_1.GtDrilldownComponent
                ],
                providers: [],
                bootstrap: []
            },] },
];
/** @nocollapse */
GenericTableModule.ctorParameters = function () { return []; };
exports.GenericTableModule = GenericTableModule;
