// @angular-generic-table/core/directives/gt-column-width.directive.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var GtColumnWidthDirective = (function () {
    function GtColumnWidthDirective(hostElement, cdRef) {
        this.hostElement = hostElement;
        this.cdRef = cdRef;
    }
    GtColumnWidthDirective.prototype.ngOnInit = function () {
        this.checkSize();
    };
    GtColumnWidthDirective.prototype.onResize = function ($event) {
        this.checkSize();
    };
    GtColumnWidthDirective.prototype.checkSize = function () {
        this.widths[this.objectKey] = window.getComputedStyle(this.hostElement.nativeElement, null).getPropertyValue('width');
        this.cdRef.detectChanges();
    };
    return GtColumnWidthDirective;
}());
GtColumnWidthDirective.decorators = [
    { type: core_1.Directive, args: [{
                selector: '[gtColumnWidth]',
                host: {
                    '(window:resize)': 'onResize($event)'
                }
            },] },
];
/** @nocollapse */
GtColumnWidthDirective.ctorParameters = function () { return [
    { type: core_1.ElementRef, },
    { type: core_1.ChangeDetectorRef, },
]; };
GtColumnWidthDirective.propDecorators = {
    'objectKey': [{ type: core_1.Input },],
    'widths': [{ type: core_1.Input },],
};
exports.GtColumnWidthDirective = GtColumnWidthDirective;
