// @angular-generic-table/core/components/gt-expanding-row.component.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var GtExpandedRow = (function () {
    function GtExpandedRow() {
        this.redrawEvent = new core_1.EventEmitter();
        this.toggleRowEvent = new core_1.EventEmitter();
    }
    GtExpandedRow.prototype.$hide = function () {
        this.toggleRowEvent.emit(this.row);
    };
    GtExpandedRow.prototype.$redraw = function () {
        this.redrawEvent.emit(this.row);
    };
    return GtExpandedRow;
}());
exports.GtExpandedRow = GtExpandedRow;
var GtExpandingRowComponent = (function () {
    function GtExpandingRowComponent() {
        this.redrawEvent = new core_1.EventEmitter();
        this.toggleRowEvent = new core_1.EventEmitter();
    }
    GtExpandingRowComponent.prototype.newInstance = function (instance) {
        instance.row = this.row;
        instance.columnWidth = this.columnWidth;
        instance.gtSettings = this.gtSettings;
        instance.gtFields = this.gtFields;
        instance.gtOptions = this.gtOptions;
        instance.gtInfo = this.gtInfo;
        instance.data = typeof this.data === 'function' ? this.data(this.row) : this.data;
        instance.redrawEvent.subscribe(this.redrawEvent);
        instance.toggleRowEvent.subscribe(this.toggleRowEvent);
    };
    return GtExpandingRowComponent;
}());
GtExpandingRowComponent.decorators = [
    { type: core_1.Component, args: [{
                selector: 'gt-expanding-row',
                template: "\n        <div appComponentAnchor\n             [ctor]=\"type\" (instance)=\"newInstance($event)\"></div>"
            },] },
];
/** @nocollapse */
GtExpandingRowComponent.ctorParameters = function () { return []; };
GtExpandingRowComponent.propDecorators = {
    'type': [{ type: core_1.Input },],
    'row': [{ type: core_1.Input },],
    'columnWidth': [{ type: core_1.Input },],
    'gtSettings': [{ type: core_1.Input },],
    'gtFields': [{ type: core_1.Input },],
    'gtOptions': [{ type: core_1.Input },],
    'gtInfo': [{ type: core_1.Input },],
    'data': [{ type: core_1.Input },],
    'redrawEvent': [{ type: core_1.Output },],
    'toggleRowEvent': [{ type: core_1.Output },],
};
exports.GtExpandingRowComponent = GtExpandingRowComponent;
